{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a74d133-10d0-4a31-91f7-675811cdcd70",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Load dependencies and custom functions\n",
    "\n",
    "from datetime import datetime\n",
    "from csv import DictReader\n",
    "from matplotlib import pyplot\n",
    "import csv\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "# This function will return a list of positions where element exists in the dataframe.\n",
    "def getIndexes(dfObj, value):\n",
    "    listOfPos = []\n",
    "    result = dfObj.isin([value])\n",
    "    seriesObj = result.any()\n",
    "    columnNames = list(seriesObj[seriesObj == True].index)\n",
    "    for col in columnNames:\n",
    "        rows = list(result[col][result[col] == True].index)\n",
    "        for row in rows:\n",
    "            listOfPos.append((row))\n",
    "    return listOfPos\n",
    "\n",
    "\n",
    "now = datetime.now()\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "print(\"Dependencies loaded at\", current_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de9dd486-f6ca-4898-8eac-a3fa44f0d499",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Specify relevant file paths and load data\n",
    "\n",
    "nucFile='/path/to/Stenstrom_NucleolarProteome.xlsx'\n",
    "idFile='/path/to/PRISM Drug Screen/TranscriptIDs.csv'\n",
    "exprFile='/path/to/CCLE_expression_filtered.csv'\n",
    "sensFile='/path/to/oxaliplatin (BRDBRD-K78960041-001-05-7) Drug sensitivity AUC (PRISM Repurposing Secondary Screen) 19Q4.csv'\n",
    "\n",
    "nucDF = pd.read_excel(nucFile, sheet_name='All nucleoli genes')\n",
    "nucGenes = nucDF.filter(items=['Gene name'])\n",
    "\n",
    "sensDF = pd.read_csv(sensFile, usecols=['Depmap ID', 'Drug sensitivity AUC (PRISM Repurposing Secondary Screen) 19Q4', 'Lineage'])\n",
    "\n",
    "now = datetime.now()\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "print(\"Data loaded at\", current_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1673e4c-3ec1-420b-a1e0-267894492d00",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Generate Transcript ID dictionary\n",
    "\n",
    "idDF = pd.DataFrame(columns = ['ID'])\n",
    "\n",
    "# open TranscriptsIDs in read mode\n",
    "with open(idFile, 'r') as read_obj:\n",
    "    id_dict_reader = DictReader(read_obj)\n",
    "    # iterate over each line in drug data file\n",
    "    for row in id_dict_reader:\n",
    "        ID = row['ID']\n",
    "        sID = ID.split(' (')\n",
    "        rID = sID[0]\n",
    "        idDF = idDF.append({'ID' : rID}, ignore_index = True)\n",
    "\n",
    "print(idDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a62a505c-e05e-48fd-9937-c477ab34b5a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "##Find nucleolar genes and calculate correlation between expression and Ox sensitivity\n",
    "\n",
    "#Iterate over all nucleolar genes from the Stenstrom dataset\n",
    "lfd = 10499\n",
    "while lfd < len(idDF):\n",
    "\n",
    "    selGene = idDF.iloc[[lfd]]\n",
    "    sele = selGene.iloc[0,0]\n",
    "    print(sele)\n",
    "\n",
    "    Pos = getIndexes(idDF, sele)\n",
    "    if len(Pos) == 1: \n",
    "        lPos = Pos[0]\n",
    "\n",
    "\n",
    "        #Load AUC data for current nucleolar gene of interest\n",
    "        filterDF = pd.read_csv(exprFile, usecols=[0,lPos+1], names=['Depmap ID', 'TPM'], header=None)\n",
    "\n",
    "        #Find intersection between expression and sensitivity data\n",
    "        mergeDF = pd.merge(filterDF, sensDF, how='inner', on='Depmap ID')\n",
    "\n",
    "\n",
    "        #Calculate correlation between drug sensitivity and expression levels across all cell lines\n",
    "        sens = mergeDF['Drug sensitivity AUC (PRISM Repurposing Secondary Screen) 19Q4']\n",
    "        expr = mergeDF['TPM']\n",
    "        corr = sens.corr(expr, method='pearson')\n",
    "        print('All:', corr)\n",
    "\n",
    "        #Filter for colorectal cancer cell lines only\n",
    "        CRC = mergeDF[mergeDF[\"Lineage\"] == \"Colorectal\"]\n",
    "        sensCRC = CRC['Drug sensitivity AUC (PRISM Repurposing Secondary Screen) 19Q4']\n",
    "        exprCRC = CRC['TPM']\n",
    "        corrCRC = sensCRC.corr(exprCRC, method='pearson')\n",
    "        print('CRC:', corrCRC)\n",
    "\n",
    "        #Filter for central nervous system cancer cell lines only\n",
    "        CNS = mergeDF[mergeDF[\"Lineage\"] == \"Central Nervous System\"]\n",
    "        sensCNS = CNS['Drug sensitivity AUC (PRISM Repurposing Secondary Screen) 19Q4']\n",
    "        exprCNS = CNS['TPM']\n",
    "        corrCNS = sensCNS.corr(exprCNS, method='pearson')\n",
    "        print('CNS:', corrCNS)\n",
    "\n",
    "        #Save correlation data \n",
    "        with open('PearsonCorrelation_AllGenes.csv', 'a') as corrfile:\n",
    "            schr = csv.writer(corrfile, quoting=csv.QUOTE_ALL)\n",
    "            schr.writerow([sele, corr, corrCRC, corrCNS])\n",
    "    \n",
    "    #Advance loop\n",
    "    print()\n",
    "    lfd += 1\n",
    "\n",
    "print(\"Done!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
